tests/TestHarness/Node.py

```diff
@@ -489,6 +490,32 @@ def waitForTransInBlock(self, transId, timeout=None):
         ret=Utils.waitForBool(lam, timeout)
         return ret
 
+    def checkBlocksForTransactions(self, transIds, lastBlockProcessed, endBlock):
+        for blockNum in range(endBlock, lastBlockProcessed + 1, -1):
+            block = self.processCurlCmd("trace_api", "get_block", f'{{"block_num":{blockNum}}}', silentErrors=False, exitOnError=True)
+            if block['transactions']:
+                for trx in block['transactions']:
+                    if trx['id'] in transIds:
+                        transIds.pop(trx['id'])
+        return transIds
+
+    def waitOnBlockTransactions(self, transIds, startBlock=1, timeout=30):
+        lastBlockProcessed = startBlock
+        finalBlock = sys.maxsize
+        if timeout is not None:
+            finalBlock = self.getHeadBlockNum() + timeout * 2
+        while len(transIds) > 0:
+            self.waitForHeadToAdvance()
+            endBlock = self.getHeadBlockNum()
+            if endBlock > finalBlock:
+                endBlock = finalBlock
+            print(lastBlockProcessed, endBlock)
+            transIds = self.checkBlocksForTransactions(transIds, lastBlockProcessed, endBlock)
```

Comment:
For this algorithm to work with basing the timeout on block production, it would appear that this function should only process one block at a time and the loop/range be managed from withing `waitOnBlockTransactions`.  This will have some benefits:

1. Can detect and manage timeout more accurately
2. Can stop querying for additional blocks in initial range if, say, all transactions were found in the first block returned from `get_block`
3. Could allow for future optimizations if empirical data suggests wanting to do a search pattern that starts checking blocks in chronological reverse order until head advances, then check the newest block before continuing with the existing blocks in reverse chronological order.

Reply:
Note: If updating to only query/process one block at a time, it will be important to add handling around the `waitForHeadToAdvance` so that it isn't always waiting when it still has blocks to process.

Reply:
I misunderstood `timeout` to be trying to constrain the amount of time spent in the waitOnBlockTransactions function call. I now understand it to be a book end on how many additional blocks this function will wait to be produced and audited for transactions before exiting and reporting not all trxs were found.